File: src/db/storage.py
import io
import logging
import sqlite3
import time
from typing import Sequence

import PIL.Image as PILImage

logger = logging.getLogger(__name__)


def store_thumbnails(
    conn: sqlite3.Connection,
    sha256: str,
    file_mime_type: str,
    process_version: int,
    thumbnails: Sequence[PILImage.Image],
):
    cursor = conn.cursor()
    # Delete existing thumbnails for the item if they have a lower version
    cursor.execute(
        """
        DELETE FROM thumbnails
        WHERE item_sha256 = ? AND version < ?
    """,
        (sha256, process_version),
    )
    cursor.executemany(
        """
    INSERT INTO thumbnails (item_sha256, idx, item_mime_type, width, height, version, thumbnail)
    VALUES (?, ?, ?, ?, ?, ?, ?)
    """,
        [
            (
                sha256,
                idx,
                file_mime_type,
                thumbnail.width,
                thumbnail.height,
                process_version,
                thumbnail_to_bytes(thumbnail, get_thumb_format(file_mime_type)),
            )
            for idx, thumbnail in enumerate(thumbnails)
        ],
    )


def get_thumb_format(item_mime_type: str) -> str:
    # Check if the source file format is lossless
    # if item_mime_type in ["image/png", "image/tiff", "image/bmp"]:
    #     return "PNG"
    # Default to JPEG for lossy formats
    return "JPEG"


def convert_image_mode(image: PILImage.Image, format: str) -> PILImage.Image:
    if format == "JPEG":
        # Allowed mode for JPEG is RGB
        if image.mode != "RGB":
            image = image.convert("RGB")
    elif format == "PNG":
        # Allowed modes for PNG are RGBA, RGB, L (grayscale), LA (grayscale with alpha)
        if image.mode not in ["RGBA", "RGB", "L", "LA"]:
            # Convert paletted images to RGBA or grayscale images to LA
            image = image.convert("RGBA")
    return image


def thumbnail_to_bytes(thumbnail: PILImage.Image, format: str = "PNG") -> bytes:
    thumbnail = convert_image_mode(thumbnail, format)
    start_time = time.time()
    with io.BytesIO() as output:
        thumbnail.save(output, format)  # Save as PNG or any other format
        size = round(output.tell() / (1024 * 1024), 4)  # Size in MB
        logger.debug(
            f"Thumbnail converted to {format} (size {size}) in {time.time() - start_time:.2f} seconds"
        )
        return output.getvalue()


def has_thumbnail(
    conn: sqlite3.Connection, sha256: str, process_version: int
) -> bool:
    cursor = conn.cursor()
    cursor.execute(
        """
        SELECT EXISTS(
            SELECT 1 
            FROM thumbnails 
            WHERE item_sha256 = ? AND idx = 0 AND version >= ?
            LIMIT 1
        )
        """,
        (sha256, process_version),
    )
    result = cursor.fetchone()
    return result[0] == 1 if result else False


def get_thumbnail(
    conn: sqlite3.Connection, sha256: str, idx: int
) -> PILImage.Image | None:
    cursor = conn.cursor()
    cursor.execute(
        """
        SELECT thumbnail 
        FROM thumbnails 
        WHERE item_sha256 = ? AND idx = ?
        """,
        (sha256, idx),
    )
    result = cursor.fetchone()
    if result:
        thumbnail_data = result[0]
        thumbnail = PILImage.open(io.BytesIO(thumbnail_data))
        return thumbnail
    else:
        return None


def delete_orphaned_thumbnails(conn: sqlite3.Connection):
    cursor = conn.cursor()
    cursor.execute(
        """
        DELETE FROM thumbnails
        WHERE item_sha256 IN (
            SELECT thumbnails.item_sha256
            FROM thumbnails
            LEFT JOIN items ON thumbnails.item_sha256 = items.sha256
            WHERE items.sha256 IS NULL
        )
        """
    )
    if cursor.rowcount > 0:
        logger.info(f"Deleted {cursor.rowcount} orphaned thumbnails")
    return cursor.rowcount  # Return the number of rows deleted


def store_frames(
    conn: sqlite3.Connection,
    sha256: str,
    file_mime_type: str,
    process_version: int,
    frames: list[PILImage.Image],
):
    cursor = conn.cursor()
    # Delete existing frames for the item if they have a lower version
    cursor.execute(
        """
        DELETE FROM frames
        WHERE item_sha256 = ? AND version < ?
    """,
        (sha256, process_version),
    )
    cursor.executemany(
        """
    INSERT INTO frames (item_sha256, idx, item_mime_type, width, height, version, frame)
    VALUES (?, ?, ?, ?, ?, ?, ?)
    """,
        [
            (
                sha256,
                idx,
                file_mime_type,
                frame.width,
                frame.height,
                process_version,
                thumbnail_to_bytes(frame, get_thumb_format(file_mime_type)),
            )
            for idx, frame in enumerate(frames)
        ],
    )


def has_frame(
    conn: sqlite3.Connection, sha256: str, process_version: int = 0
) -> bool:
    cursor = conn.cursor()
    cursor.execute(
        """
        SELECT EXISTS(
            SELECT 1 
            FROM frames 
            WHERE item_sha256 = ? AND idx = 0 AND version >= ?
            LIMIT 1
        )
        """,
        (sha256, process_version),
    )
    result = cursor.fetchone()
    return result[0] == 1 if result else False


def get_frames(conn: sqlite3.Connection, sha256: str) -> list[PILImage.Image]:
    cursor = conn.cursor()
    cursor.execute(
        """
        SELECT frame 
        FROM frames 
        WHERE item_sha256 = ?
        ORDER BY idx
        """,
        (sha256,),
    )
    results = cursor.fetchall()
    frames = []
    for result in results:
        frame_data = result[0]
        frame = PILImage.open(io.BytesIO(frame_data))
        frames.append(frame)
    return frames


def delete_orphaned_frames(conn: sqlite3.Connection):
    cursor = conn.cursor()
    cursor.execute(
        """
        DELETE FROM frames
        WHERE item_sha256 IN (
            SELECT frames.item_sha256
            FROM frames
            LEFT JOIN items ON frames.item_sha256 = items.sha256
            WHERE items.sha256 IS NULL
        )
        """
    )
    if cursor.rowcount > 0:
        logger.info(f"Deleted {cursor.rowcount} orphaned frames")
    return cursor.rowcount  # Return the number of rows deleted

File: gateway/src/db/storage.rs
use crate::api_error::ApiError;

type ApiResult<T> = std::result::Result<T, ApiError>;

pub(crate) struct StoredImage {
    pub idx: i64,
    pub width: i64,
    pub height: i64,
    pub bytes: Vec<u8>,
}

pub(crate) async fn has_thumbnail(
    conn: &mut sqlx::SqliteConnection,
    sha256: &str,
    process_version: i64,
) -> ApiResult<bool> {
    let row: (i64,) = sqlx::query_as(
        r#"
SELECT EXISTS(
    SELECT 1
    FROM storage.thumbnails
    WHERE item_sha256 = ?1 AND idx = 0 AND version >= ?2
    LIMIT 1
) AS exists_flag
        "#,
    )
    .bind(sha256)
    .bind(process_version)
    .fetch_one(&mut *conn)
    .await
    .map_err(|err| {
        tracing::error!(error = %err, "failed to check thumbnail existence");
        ApiError::internal("Failed to read thumbnail")
    })?;

    Ok(row.0 == 1)
}

pub(crate) async fn has_frame(
    conn: &mut sqlx::SqliteConnection,
    sha256: &str,
    process_version: i64,
) -> ApiResult<bool> {
    let row: (i64,) = sqlx::query_as(
        r#"
SELECT EXISTS(
    SELECT 1
    FROM storage.frames
    WHERE item_sha256 = ?1 AND idx = 0 AND version >= ?2
    LIMIT 1
) AS exists_flag
        "#,
    )
    .bind(sha256)
    .bind(process_version)
    .fetch_one(&mut *conn)
    .await
    .map_err(|err| {
        tracing::error!(error = %err, "failed to check frame existence");
        ApiError::internal("Failed to read frame")
    })?;

    Ok(row.0 == 1)
}

pub(crate) async fn store_thumbnails(
    conn: &mut sqlx::SqliteConnection,
    sha256: &str,
    mime_type: &str,
    process_version: i64,
    thumbnails: &[StoredImage],
) -> ApiResult<()> {
    sqlx::query(
        r#"
DELETE FROM storage.thumbnails
WHERE item_sha256 = ?1 AND version < ?2
        "#,
    )
    .bind(sha256)
    .bind(process_version)
    .execute(&mut *conn)
    .await
    .map_err(|err| {
        tracing::error!(error = %err, "failed to prune thumbnails");
        ApiError::internal("Failed to store thumbnails")
    })?;

    for thumb in thumbnails {
        sqlx::query(
            r#"
INSERT INTO storage.thumbnails (
    item_sha256, idx, item_mime_type, width, height, version, thumbnail
)
VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7)
            "#,
        )
        .bind(sha256)
        .bind(thumb.idx)
        .bind(mime_type)
        .bind(thumb.width)
        .bind(thumb.height)
        .bind(process_version)
        .bind(&thumb.bytes)
        .execute(&mut *conn)
        .await
        .map_err(|err| {
            tracing::error!(error = %err, "failed to store thumbnail");
            ApiError::internal("Failed to store thumbnails")
        })?;
    }

    Ok(())
}

pub(crate) async fn store_frames(
    conn: &mut sqlx::SqliteConnection,
    sha256: &str,
    mime_type: &str,
    process_version: i64,
    frames: &[StoredImage],
) -> ApiResult<()> {
    sqlx::query(
        r#"
DELETE FROM storage.frames
WHERE item_sha256 = ?1 AND version < ?2
        "#,
    )
    .bind(sha256)
    .bind(process_version)
    .execute(&mut *conn)
    .await
    .map_err(|err| {
        tracing::error!(error = %err, "failed to prune frames");
        ApiError::internal("Failed to store frames")
    })?;

    for frame in frames {
        sqlx::query(
            r#"
INSERT INTO storage.frames (
    item_sha256, idx, item_mime_type, width, height, version, frame
)
VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7)
            "#,
        )
        .bind(sha256)
        .bind(frame.idx)
        .bind(mime_type)
        .bind(frame.width)
        .bind(frame.height)
        .bind(process_version)
        .bind(&frame.bytes)
        .execute(&mut *conn)
        .await
        .map_err(|err| {
            tracing::error!(error = %err, "failed to store frame");
            ApiError::internal("Failed to store frames")
        })?;
    }

    Ok(())
}

pub(crate) async fn delete_orphaned_thumbnails(
    conn: &mut sqlx::SqliteConnection,
) -> ApiResult<u64> {
    let result = sqlx::query(
        r#"
DELETE FROM storage.thumbnails
WHERE item_sha256 IN (
    SELECT storage.thumbnails.item_sha256
    FROM storage.thumbnails
    LEFT JOIN items ON storage.thumbnails.item_sha256 = items.sha256
    WHERE items.sha256 IS NULL
)
        "#,
    )
    .execute(&mut *conn)
    .await
    .map_err(|err| {
        tracing::error!(error = %err, "failed to delete orphaned thumbnails");
        ApiError::internal("Failed to delete orphaned thumbnails")
    })?;

    Ok(result.rows_affected())
}

pub(crate) async fn delete_orphaned_frames(
    conn: &mut sqlx::SqliteConnection,
) -> ApiResult<u64> {
    let result = sqlx::query(
        r#"
DELETE FROM storage.frames
WHERE item_sha256 IN (
    SELECT storage.frames.item_sha256
    FROM storage.frames
    LEFT JOIN items ON storage.frames.item_sha256 = items.sha256
    WHERE items.sha256 IS NULL
)
        "#,
    )
    .execute(&mut *conn)
    .await
    .map_err(|err| {
        tracing::error!(error = %err, "failed to delete orphaned frames");
        ApiError::internal("Failed to delete orphaned frames")
    })?;

    Ok(result.rows_affected())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::db::migrations::setup_test_databases;

    // Ensures storage cleanup removes thumbnails that no longer have corresponding items.
    #[tokio::test]
    async fn delete_orphaned_thumbnails_removes_missing_items() {
        let mut dbs = setup_test_databases().await;
        sqlx::query(
            r#"
INSERT INTO items (id, sha256, md5, type, time_added)
VALUES (1, 'sha_one', 'md5_one', 'image/png', '2024-01-01T00:00:00')
            "#,
        )
        .execute(&mut dbs.index_conn)
        .await
        .unwrap();
        sqlx::query(
            r#"
INSERT INTO storage.thumbnails (item_sha256, idx, item_mime_type, width, height, version, thumbnail)
VALUES
    ('sha_one', 0, 'image/png', 10, 10, 1, x'00'),
    ('sha_missing', 0, 'image/png', 10, 10, 1, x'00')
            "#,
        )
        .execute(&mut dbs.index_conn)
        .await
        .unwrap();

        let deleted = delete_orphaned_thumbnails(&mut dbs.index_conn)
            .await
            .unwrap();
        assert_eq!(deleted, 1);
    }

    // Ensures storage cleanup removes frames that no longer have corresponding items.
    #[tokio::test]
    async fn delete_orphaned_frames_removes_missing_items() {
        let mut dbs = setup_test_databases().await;
        sqlx::query(
            r#"
INSERT INTO items (id, sha256, md5, type, time_added)
VALUES (1, 'sha_one', 'md5_one', 'image/png', '2024-01-01T00:00:00')
            "#,
        )
        .execute(&mut dbs.index_conn)
        .await
        .unwrap();
        sqlx::query(
            r#"
INSERT INTO storage.frames (item_sha256, idx, item_mime_type, width, height, version, frame)
VALUES
    ('sha_one', 0, 'image/png', 10, 10, 1, x'00'),
    ('sha_missing', 0, 'image/png', 10, 10, 1, x'00')
            "#,
        )
        .execute(&mut dbs.index_conn)
        .await
        .unwrap();

        let deleted = delete_orphaned_frames(&mut dbs.index_conn).await.unwrap();
        assert_eq!(deleted, 1);
    }
}
